### PR TITLE
Implement Pagination in the GET Opportunities API at desired location

### LINKED ISSUE #44: Implement Pagination in the GET Opportunities API
1. Use two additional query params - `limit` and `offset` 
Instead of returning all the opportunities return something like: limited_query_set = query_set[offset : offset + limit]. 
An Example: if offset = 7 & limit = 20, The opportunities 7th to 27th should be returned.  
PS:
Use
"DEFAULT_PAGE_LIMIT" - 20
"DEFAULT_PAGE_OFFSET", - 0
The changes will be effectively done in `/managers/opportunity/index.js` in the `getOpportunities` function.

2. Update the API documentation accordingly inside `routes/opportunity.js`

### PR DESCRIPTION
**Pull request by Kashish0423**
Issue no 44

Implemented Pagination as per the Project Requirement ,
Commented all the functionality in the code itself .
please have a look at it whether or not it is okay with the project , if yes then I will further update the Documentation as well  and do suggest necessary changes if any

### CODE REVIEW

File: managers/opportunity/index.js [hunk 1]
```diff
@@ -48,16 +59,47 @@ class opportunityManager {
...
-      return fetchedOpportunitiesQuery;
+      if(queryLimit==undefined || queryLimit<0 || offset< 0)
+      {
+        queryLimit= default_page_limit ;
+      }
+
+      if(offset==undefined || offset<0 || queryLimit<0)
+      {
+        offset= default_page_offset ;
+      }
```
Comment (abdus) [inline managers/opportunity/index.js]: - what if I pass a huge value to `quaryLimit`? how does that will work? will it slow down DB operation or crash it completely? 
- I think these `if` conditions are not correct. should the value of `offset` affect `queryLimit`? 
- let the user know when there is a 'next page' available. for that, you have to modify the output format a bit.

**Review by abdus — CHANGES_REQUESTED** [review]

File: managers/opportunity/index.js (hunk 1 above)
Comment (Kashish0423) [reply managers/opportunity/index.js]: okay as per this :   
1 .   I haven't worked with huge databases but as far as queryLimit being very high , I can see that this is an O(limit) operation
as it will try and get all the data , in case of data sets above 10^8/9  it will not work in the time limit provide, but I don't think we can improve on this as we aren't performing any operation on the range (else we could have used Segment/ Fenwick trees to make each query O(Logn) compared to O(n) which is right now ), 
a Preventive measure that I can think of in real world cases will be having multiple servers / instances to handle very big amount of queries (ie 10^5 or more ) or in else in normal case it will work fine as queries are very limited  

2. Offset here denotes the starting position from where I Start fetching the data , it will affect the limit in those cases when  (offset+ limit > dataset size ) for that we can make a check on limit tot bring it down to (dataSetsize- offset), I m not sure about exact behaviour of mongoose limit but I m guessing it does that by default  and if negative or undefined values are considered , it is already handled . 

3.  okay , I will have to think around this a bit but most probably it will be something like keeping tracking of current position of the user in the database and if the user is on the last page (then return something like no more pages ahead  or page limit reached ) and in other cases just simply make a next page available option valid ,

But I have a doubt what is next page exactly here as we are dealing in terms of database index or is it some sort of other page ?

Comment (abdus) [reply managers/opportunity/index.js]: >  I haven't worked with huge databases but as far as queryLimit being very high ... 

my concern was about the user's input. what if the user intentionally puts huge numbers to trick the system? I think we could set a maximum offset. 

> But I have a doubt what is next page exactly here .... 

for example, if the offset is 0 and limit is 10, and there are more than 30 entries in the database, we should somehow let the client know that there are more data that could be fetched on subsequent request. 

Following is a snippet from a codebase I am working on, where I have implemented pagination. Go through it and see if you could find something valuable. If you find any issue with this snippet, please don't hesitate to point it out. 

```typescript
const RESULT_PER_PAGE = 10;

interface IOptions {
  page?: number;
}

async function joinFeedAndItem(
  options?: IOptions
): Promise<{ data: any[]; page: number; nextPage: number | null }> {
  const page = getValidPageNum(options?.page);
  const response = [];
  const feedItems = await FeedItemModel.find()
    .skip(RESULT_PER_PAGE * page)
    .limit(RESULT_PER_PAGE)
    .exec();

  const nextPage: number | null =
    (await FeedItemModel.countDocuments()) >
    RESULT_PER_PAGE * page + RESULT_PER_PAGE
      ? page + 1
      : null;

  for (let i = 0; i < feedItems.length; i++) {
    const fItem = feedItems[i].toObject();
    const feed = (
      await FeedModel.findOne({
        item: fItem._id,
      })
    ).toObject();

    (fItem as any).feed = feed || 'feed missing';
    (fItem as any).feed.itemCount = (feed as any).item?.length;

    delete (feed as any).item;
    response.push(fItem);
  }

  return { data: response, page, nextPage };
}

function getValidPageNum(num: number) {
  return (!Number.isNaN(num) && Math.max(Math.min(1000, num), 0)) || 0;
}
```

Comment (Kashish0423) [reply managers/opportunity/index.js]: Okay , I get your point now, 
 So let's suppose the user tries to trick the database, for example let's consider a DB with size 100 , user sets offset as 50 and limit as 100 , here we have following options :

1: Return a list with all the entries with 50 to 100
2 : don't do anything , just return an error that it is an invalid request 
3 : we can return a list of all the entries from 50 to 100 along with a message that no more pages can be fetched and 
    also in our server turning his request size from 100 to 50 only (max_size- offset)  always and keeping a conditional 
    (limit <=max_size- offset ? valid : false  && (offset validated before by either making it default or returning an error )  ) 
  to  make sure it doesn't exceed ever 
4 : For handle out of bound queries by just setting it to some default , but it has a downside that the server is always busy 
   in fetching down data even for invalid queries 

As of now I can think of all above options that can be implemented depending upon various scenarios ,

Yes, your above code does help me in understanding the the implementation of getting the next page but can you tell me which option to go for of the ones I  mentioned or any other if there to handle invalid/tricky query parameters

Comment (abdus) [reply managers/opportunity/index.js]: respond with a valid error message. that should be enough.

Comment (Kashish0423) [reply managers/opportunity/index.js]: okay

Comment (Kashish0423) [reply managers/opportunity/index.js]: Hello @abdus sir, sorry for the delay in the nex PR , as currently I am figuring out few things that should be known before implementing this properly , some new concepts discovered, so I just need a few extra days to get it done.

Comment (Kashish0423) [reply managers/opportunity/index.js]: Hello @abdus @Manvityagi  Please check the Updated Code of Get Opportunities Function, 

Changes Made :

1. Now the function  tells whether there is a previous  and next Page available 
2. It is able to handle Out of Bound Queries 
3. For default I have set Page=1 and Limit=1 

Sorry for the Delay, I was having issues with the concept of async wait and how to wait , so I had to go through it a bit , therefore it took time 

Please review it and provided necessary Test cases / Edits if any , I will happy to get them this time around as soon as Possible

File: managers/opportunity/index.js [hunk 2]
```diff
@@ -39,20 +39,134 @@ class opportunityManager {
     }
   }
 
+  // http://localhost:3030/opportunity?limit=4&page=2, Example of how to use the Get Method
+
+  //  page => Means the Current Page that the User Wants
+  //  limit => The count of pages that should be rendered starting from the given page
```
Comment (Manvityagi) [inline managers/opportunity/index.js:45]: limit => The count of pages that should be rendered starting from the given page

limit => The count of opportunities that should be rendered starting from the first opportunity on given page

File: managers/opportunity/index.js [hunk 3]
```diff
@@ -39,20 +39,134 @@ class opportunityManager {
...
+
+  //  For the above specified example , we Get Pages 2,3,4,5 (having Opprtunity_id as 1,2,3,4 and respectively) .
```
Comment (Manvityagi) [inline managers/opportunity/index.js:47]: If page = 2 and limit is 4, 
then we should get opportunities with id `4,5,6,7` (assuming ids are 0,1,2,3,4,5,6,7, etc) , Isn't it? 
@Kashish0423

**Review by Manvityagi — CHANGES_REQUESTED** [review]
Assuming Opportunity Ids are (0,1,2,3,4....) 

as per your code, 

page=1 & limit = 1 returns opportunities 0,1 (which is fine) 
page=2 & limit = 2 returns opportunities 1,2 (which is not fine, it should have returned 2,3) 
page=3 & limit = 2 returns opportunities 2,3 (which is not fine, it should have returned 4,5)  

I think there is some confusion in the concept of page and limit, 
Let me clear it once again, 
Limit means how many opportunities are at 1 page, 
Page means, for which page we are requesting the opportunities. 

Some examples: 
For limit 3, 
Page 1 - 0,1,2
Page 2 - 3, 4, 5
Page 3 - 6, 7, 8
Page 4 - 9,10 (if not in multiple of 3, last page should have whatever left) 

For limit 4, 
Page 1 - 0,1,2,3
Page 2 - 4,5,6,7
Page 3 - 8,9,10,11

I hope we are on the same page?

**Comment by Manvityagi** [comment]
I think it has been my mistake, I conveyed the meaning of offset wrongly in the beginning at the time issue was opened, 
but now I think with my [previous comment](https://github.com/Girl-Code-It/Opportunity-Calendar-Backend/pull/47#pullrequestreview-640276823), expectations are clear? 

To compensate for the extra changes that you have to do because of our mistake, I can increase the level of this issue by 1. 
Is that okay with you @Kashish0423. 

And once again sorry for the miscommunication.

CC: @abdus

File: managers/opportunity/index.js (hunk 3 above)
Comment (Kashish0423) [reply managers/opportunity/index.js:47]: Hello @Manvityagi  mam sorry for the late response , now  I get your point about the page and the limit ,  with the example you explained above , I will try to get it working as soon as possible . Is the level increased of the issue ? It's still showing me level 2 at the moment .

Comment (Manvityagi) [reply managers/opportunity/index.js:47]: @Kashish0423 Yes it has been changed to level 3, Waiting for your PR.

Comment (Kashish0423) [reply managers/opportunity/index.js:47]: Yes Mam , on it and  trying to get it done asap

**Comment by Kashish0423** [comment]
Hello @Manvityagi  mam please have a look I have implemented the Pagination as requested 

1. There  is one issue, I don't know where to change/ see the actual Database but there is an opportunity with opportunity_id as 12 but it is placed after 6 , although there is no problem in rendering but it might create confusion , so should be changed to 7 in my opinion for more easy understanding of the code . 

2. Rest, if any test cases fail do let me know I will be happy to get it done . Also, I have tried the ones that I could think of , considering almost all the corner cases like before

**Comment by Manvityagi** [comment]
> Hello @Manvityagi mam please have a look I have implemented the Pagination as requested
> 
> 1. There is one issue, I don't know where to change/ see the actual Database but there is an opportunity with opportunity_id as 12 but it is placed after 6 , although there is no problem in rendering but it might create confusion , so should be changed to 7 in my opinion for more easy understanding of the code .
> 2. Rest, if any test cases fail do let me know I will be happy to get it done . Also, I have tried the ones that I could think of , considering almost all the corner cases like before

1. It's okay, the opportunity_id field will be soon deleted from the schema anyway. 
2. To see the database, you can install mongo compass, and copy the connections string(the credentials are in .env.example file, which you also must have copied in a .env file), and while the string is copied, open mongo compass, it will be automatically detected, enter the password and you will be able to access the dev database.

**Review by Manvityagi — CHANGES_REQUESTED** [review]
The PR has merge conflicts which should be resolved before merging, Please resolve them.

**Comment by Kashish0423** [comment]
#68 Please have a look at it here @Manvityagi  mam (backend_pagination branch ), I have implemented the same logic but just used another branch , as I was a little confused about how to resolve conflicts  in my normal branch which I was using earlier , but in the meanwhile I have resolved conflicts using the that Branch(getOpportunity_branch)  as but don't know where to make a PR  as it is showing close, please let me know how what to do to complete the issue .

**Comment by Manvityagi** [comment]
> #68 Please have a look at it here @Manvityagi mam (backend_pagination branch ), I have implemented the same logic but just used another branch , as I was a little confused about how to resolve conflicts in my normal branch which I was using earlier , but in the meanwhile I have resolved conflicts using the that Branch(getOpportunity_branch) as but don't know where to make a PR as it is showing close, please let me know how what to do to complete the issue .

You have made a new PR #68, proceed with it.
